ompi/mca/coll/han/coll_han_alltoall.c

```diff
@@ -207,8 +207,8 @@ int mca_coll_han_alltoall_using_smsc(
             /* all ranks will pull from the other ranks' sbuf */
             gather_buf_in[0] = (void*)sbuf;
         }
-        gather_buf_in[1] = *(void**)&send_needs_bounce;
-        gather_buf_in[2] = *(void**)&ii_push_data;
+        gather_buf_in[1] = (void*)&send_needs_bounce;
+        gather_buf_in[2] = (void*)&ii_push_data;
```

Comment:
This isn't the same. The result of the original lines was the values in `send_needs_bounce` and `ii_push_data` cast to `void*`. The new lines result in the pointer to these variables. I suspect that the new behavior is not what was intended (since it's passed to inside an array to `allgather`).

Reply:
Ok, gotcha.  Should they then be:
```c
gather_buf_in[1] = (void*)send_needs_bounce;
gather_buf_in[2] = (void*)ii_push_data;
```

Reply:
Yes, that is probably what was intended. Maybe @lrbison should chime in though :)

Reply:
@lrbison I actually ended up changing it to:

```diff
diff --git a/ompi/mca/coll/han/coll_han_alltoall.c b/ompi/mca/coll/han/coll_han_alltoall.c
index c2f4b58e32..8c1a601c54 100644
--- a/ompi/mca/coll/han/coll_han_alltoall.c
+++ b/ompi/mca/coll/han/coll_han_alltoall.c
@@ -207,8 +207,8 @@ start_allgather:
             /* all ranks will pull from the other ranks' sbuf */
             gather_buf_in[0] = (void*)sbuf;
         }
-        gather_buf_in[1] = *(void**)&send_needs_bounce;
-        gather_buf_in[2] = *(void**)&ii_push_data;
+        gather_buf_in[1] = (void*)(intptr_t)send_needs_bounce;
+        gather_buf_in[2] = (void*)(intptr_t)ii_push_data;
 
         rc = low_comm->c_coll->coll_allgather(gather_buf_in, nptrs_gather, MPI_AINT,
                     gather_buf_out, nptrs_gather, MPI_AINT, low_comm,
```

Is that what was intended?

How do you extract these int values out of `gather_buf_in`?

Reply:
Thank you Joseph for catching that!  My eyeball-compilers lied and said it was the same.  The intent is to perform one gather with several values: one pointer, and two boolean flags.  Rather than make a datatype to capture the three values, I just shoved the flags each into their own pointer slot.

The compiler really is finding a potential bug here.  The code as-written leaves some high-order bits undefined.  Luckily I masked with 0x1 when reading the value, so it works as intended.

Jeff: thank you, that is the right fix.  Casting it to intptr_t makes it an int with the same size of void* and then it can be safely cast as expected and without warning.

For reference, here is where I used the value later on: 

https://github.com/open-mpi/ompi/blob/main/ompi/mca/coll/han/coll_han_alltoall.c#L224